django/db/migrations/state.py

```diff
@@ -188,6 +193,89 @@ def _reload(self, related_models):
         # Render all models
         self.apps.render_multiple(states_to_be_rendered)
 
+    def resolve_fields_and_relations(self):
+        for model_state in self.models.values():
+            for field_name, field in model_state.fields:
+                field.name = field_name
+                field.model = "{}.{}".format(model_state.app_label, model_state.name)
+                if field.remote_field and field.remote_field.model:
+                    self.fill_remote_field_name(field)
+
+        self.relations = defaultdict(lambda: defaultdict(list))
+        concretes, proxies = self._get_concrete_models_mapping_and_proxy_models()
+
+        real_apps = set(self.real_apps)
+        for model_key in concretes:
+            model_state = self.models[model_key]
+            for field_name, field in model_state.fields:
+                remote_field = field.remote_field
+                if remote_field:
+                    remote_model_key = resolve_model_key(*model_key, remote_field.model)
+                    if remote_model_key[0] not in real_apps and remote_model_key in concretes:
+                        remote_model_key = concretes[remote_model_key]
+                    self.relations[remote_model_key][model_key].append((field_name, field))
+
+                    through = getattr(remote_field, 'through', None)
+                    if through and not model_state.options.get('auto_created'):
+                        through_model_key = resolve_model_key(*model_key, through)
+                        if through_model_key[0] not in real_apps:
+                            through_model_key = concretes[through_model_key]
+                        self.relations[through_model_key][model_key].append((field_name, field))
+        for model_key in proxies:
+            self.relations[model_key] = self.relations[concretes[model_key]]
+
+    def fill_remote_field_name(self, field):
```

Comment:
I wonder if this is something we could get rid of, `RenameField.state_forwards` has to do a bit of gymnastic to handle that as well. Do you remember what breaks when `field_name` is not assigned properly?

Reply:
Not much would fail on the autodetector logic actually. It is more that multiple tests are currently asserting that the remote field name. See https://github.com/django/django/pull/12688/files#diff-c11e6432df7086eda3dfb9ab8e5b2839R1691 or https://github.com/django/django/pull/12688/files#diff-c11e6432df7086eda3dfb9ab8e5b2839R1790 among others (6 tests are failing).

If you think that these tests can be removed or substituted, that would definitely simplify the logic and remove code. However, I wouldn't know how to meaningfully replace the `field_name` asserts in the `test_unmanaged_custom_pk` test for instance (https://github.com/django/django/blob/master/tests/migrations/test_autodetector.py#L1735-L1747)